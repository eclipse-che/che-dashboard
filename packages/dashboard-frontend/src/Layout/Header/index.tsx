/*
 * Copyright (c) 2018-2023 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { PageHeader } from '@patternfly/react-core';
import { History } from 'history';
import React from 'react';
import HeaderTools from './Tools';

type Props = {
  history: History;
  isVisible: boolean;
  logo: React.ReactNode;
  logout: () => void;
  toggleNav: () => void;
};
type State = {
  isVisible: boolean;
};

export default class Header extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      isVisible: this.props.isVisible,
    };
  }

  private toggleNav(): void {
    this.props.toggleNav();
  }

  public componentDidUpdate(prevProps: Props): void {
    if (prevProps.isVisible !== this.props.isVisible) {
      this.setState({
        isVisible: this.props.isVisible,
      });
    }
  }

  public render(): React.ReactElement {
    const className = this.state.isVisible ? 'show-header' : 'hide-header';

    return (
      <PageHeader
        style={{ zIndex: 'inherit' }}
        className={className}
        logoComponent="div"
        logo={this.props.logo}
        showNavToggle={true}
        onNavToggle={() => this.toggleNav()}
        headerTools={
          <HeaderTools history={this.props.history} logout={() => this.props.logout()} />
        }
      />
    );
  }
}
