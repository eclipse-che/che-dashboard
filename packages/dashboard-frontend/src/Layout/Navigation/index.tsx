/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Nav } from '@patternfly/react-core';
import { History, Location, UnregisterCallback } from 'history';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import NavigationMainList from '@/Layout/Navigation/MainList';
import NavigationRecentList from '@/Layout/Navigation/RecentList';
import { CREATE_NEW_IF_EXIST_SWITCH_ID } from '@/pages/GetStarted/SamplesList/Toolbar/CreateNewIfExistSwitch';
import { TEMPORARY_STORAGE_SWITCH_ID } from '@/pages/GetStarted/SamplesList/Toolbar/TemporaryStorageSwitch';
import { ROUTE } from '@/Routes';
import { buildGettingStartedLocation, buildWorkspacesLocation } from '@/services/helpers/location';
import { Workspace } from '@/services/workspace-adapter';
import { RootState } from '@/store';
import { selectAllWorkspaces, selectRecentWorkspaces } from '@/store/Workspaces/selectors';

export interface NavigationItemObject {
  to: string;
  label: string;
  icon?: React.ReactElement;
}
export interface NavigationRecentItemObject {
  to: string;
  label: string;
  workspace: Workspace;
}

type Props = MappedProps & {
  history: History;
};

type State = {
  activeLocation: Location;
};

export class Navigation extends React.PureComponent<Props, State> {
  private static createNew: { isChecked: boolean | undefined } = { isChecked: undefined };
  private static temporaryStorage: { isChecked: boolean | undefined } = { isChecked: undefined };
  private static pageStateCallbacks: {
    [CREATE_NEW_IF_EXIST_SWITCH_ID]: ((isChecked: boolean | undefined) => void)[];
    [TEMPORARY_STORAGE_SWITCH_ID]: ((isChecked: boolean | undefined) => void)[];
  } = {
    [CREATE_NEW_IF_EXIST_SWITCH_ID]: [],
    [TEMPORARY_STORAGE_SWITCH_ID]: [],
  };
  // Static page state to store the state of switches across the application
  public static pageState: {
    [CREATE_NEW_IF_EXIST_SWITCH_ID]: { isChecked: boolean | undefined };
    [TEMPORARY_STORAGE_SWITCH_ID]: { isChecked: boolean | undefined };
    subscribe: (callback: (isChecked: boolean | undefined) => void, id: string) => void;
    unsubscribe: (callback: (isChecked: boolean | undefined) => void, id: string) => void;
  } = Object.freeze({
    get [CREATE_NEW_IF_EXIST_SWITCH_ID](): { isChecked: boolean | undefined } {
      return Navigation.createNew;
    },
    set [CREATE_NEW_IF_EXIST_SWITCH_ID](value: { isChecked: boolean | undefined }) {
      Navigation.createNew = value;
      Navigation.pageStateCallbacks[CREATE_NEW_IF_EXIST_SWITCH_ID].forEach(callback => {
        callback(value.isChecked);
      });
    },
    get [TEMPORARY_STORAGE_SWITCH_ID](): { isChecked: boolean | undefined } {
      return Navigation.temporaryStorage;
    },
    set [TEMPORARY_STORAGE_SWITCH_ID](value: { isChecked: boolean | undefined }) {
      Navigation.temporaryStorage = value;
      Navigation.pageStateCallbacks[TEMPORARY_STORAGE_SWITCH_ID].forEach(callback =>
        callback(value.isChecked),
      );
    },
    subscribe: (callback: (isChecked: boolean | undefined) => void, id: string): void => {
      if (Navigation.pageStateCallbacks[id]) {
        Navigation.pageStateCallbacks[id].push(callback);
      }
    },
    unsubscribe: (callback: (isChecked: boolean | undefined) => void, id: string): void => {
      if (Navigation.pageStateCallbacks[id]) {
        Navigation.pageStateCallbacks[id] = Navigation.pageStateCallbacks[id].filter(
          (cb: (isChecked: boolean | undefined) => void) => cb !== callback,
        );
      }
    },
  });
  private readonly unregisterFn: UnregisterCallback;

  constructor(props: Props) {
    super(props);

    const activeLocation = this.props.history.location;
    let newLocation: Location | undefined;

    if (activeLocation.pathname === ROUTE.HOME) {
      const workspacesNumber = this.props.allWorkspaces.length;
      if (workspacesNumber === 0) {
        newLocation = buildGettingStartedLocation();
      } else {
        newLocation = buildWorkspacesLocation();
      }
    }
    if (newLocation) {
      this.props.history.replace(newLocation);
    }

    this.state = {
      activeLocation: newLocation || activeLocation,
    };

    this.unregisterFn = this.props.history.listen((location: Location) => {
      this.setActivePath(location.pathname);
    });
  }

  private handleNavSelect(selected: {
    groupId: string | number;
    itemId: string | number;
    to: string;
    event: React.FormEvent<HTMLInputElement>;
  }): void {
    const activeLocation = {
      pathname: selected.itemId,
    } as Location;
    this.setState({
      activeLocation,
    });
  }

  private setActivePath(path: string): void {
    const activeLocation = {
      pathname: path,
    } as Location;
    this.setState({
      activeLocation,
    });
  }

  public componentWillUnmount(): void {
    this.unregisterFn();
  }

  public render(): React.ReactElement {
    const { recentWorkspaces } = this.props;
    const { activeLocation } = this.state;

    return (
      <Nav aria-label="Navigation" onSelect={selected => this.handleNavSelect(selected)}>
        <NavigationMainList activePath={activeLocation.pathname} />
        <NavigationRecentList activePath={activeLocation.pathname} workspaces={recentWorkspaces} />
      </Nav>
    );
  }
}

const mapStateToProps = (state: RootState) => ({
  recentWorkspaces: selectRecentWorkspaces(state),
  allWorkspaces: selectAllWorkspaces(state),
});
const connector = connect(mapStateToProps);

type MappedProps = ConnectedProps<typeof connector>;

export default connector(Navigation);
