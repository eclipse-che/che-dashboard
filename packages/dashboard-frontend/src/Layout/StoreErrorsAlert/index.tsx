/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { AlertVariant } from '@patternfly/react-core';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import { lazyInject } from '@/inversify.config';
import { AppAlerts } from '@/services/alerts/appAlerts';
import { AppState } from '@/store';
import { selectRegistriesErrors } from '@/store/DevfileRegistries/selectors';
import { selectInfrastructureNamespacesError } from '@/store/InfrastructureNamespaces/selectors';
import { selectPluginsError } from '@/store/Plugins/chePlugins/selectors';
import { selectDwDefaultEditorError } from '@/store/Plugins/devWorkspacePlugins/selectors';
import { selectSanityCheckError } from '@/store/SanityCheck/selectors';
import { selectUserProfileError } from '@/store/User/Profile/selectors';
import { selectWorkspacesError } from '@/store/Workspaces/selectors';

type Props = MappedProps;

export class StoreErrorsAlert extends React.PureComponent<Props> {
  @lazyInject(AppAlerts)
  private readonly appAlerts: AppAlerts;

  public componentDidUpdate(prevProps: Props) {
    // sanity check error
    if (this.props.sanityCheckError && prevProps.sanityCheckError === undefined) {
      this.appAlerts.showAlert({
        key: 'authorization-error',
        title: this.props.sanityCheckError,
        variant: AlertVariant.danger,
        timeout: false,
      });
      // do not show other errors if any
      return;
    }
  }

  public componentDidMount(): void {
    // sanity check error
    if (this.props.sanityCheckError) {
      this.appAlerts.showAlert({
        key: 'authorization-error',
        title: this.props.sanityCheckError,
        variant: AlertVariant.danger,
        timeout: false,
      });
      // do not show other errors if any
      return;
    }

    // workspaces errors
    if (this.props.workspacesError) {
      this.appAlerts.showAlert({
        key: 'workspaces-error',
        title: this.props.workspacesError,
        variant: AlertVariant.danger,
      });
    }
    // devfile registries errors
    if (this.props.registriesErrors.length > 0) {
      this.props.registriesErrors.forEach(error => {
        this.appAlerts.showAlert({
          key: 'registry-error-' + error.url,
          title: error.errorMessage,
          variant: AlertVariant.danger,
        });
      });
    }
    // plugins error
    if (this.props.pluginsError) {
      this.appAlerts.showAlert({
        key: 'plugins-error',
        title: this.props.pluginsError,
        variant: AlertVariant.danger,
      });
    }
    // devWorkspace default editor error
    if (this.props.dwDefaultEditorError) {
      this.appAlerts.showAlert({
        key: 'dw-plugins-error',
        title: this.props.dwDefaultEditorError,
        variant: AlertVariant.danger,
      });
    }
    // infrastructure namespace error
    if (this.props.infrastructureNamespacesError) {
      this.appAlerts.showAlert({
        key: 'infrastructure-namespaces-error',
        title: this.props.infrastructureNamespacesError,
        variant: AlertVariant.danger,
      });
    }
    // user profile error
    if (this.props.userProfileError) {
      this.appAlerts.showAlert({
        key: 'user-profile-error',
        title: this.props.userProfileError,
        variant: AlertVariant.danger,
      });
    }
  }

  render() {
    return '';
  }
}

const mapStateToProps = (state: AppState) => ({
  registriesErrors: selectRegistriesErrors(state),
  pluginsError: selectPluginsError(state),
  sanityCheckError: selectSanityCheckError(state),
  dwDefaultEditorError: selectDwDefaultEditorError(state),
  infrastructureNamespacesError: selectInfrastructureNamespacesError(state),
  userProfileError: selectUserProfileError(state),
  workspacesError: selectWorkspacesError(state),
});

const connector = connect(mapStateToProps);

type MappedProps = ConnectedProps<typeof connector>;
export default connector(StoreErrorsAlert);
