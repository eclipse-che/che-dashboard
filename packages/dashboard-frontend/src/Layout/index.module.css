.cheBrand {
  min-height: 50px;
}
