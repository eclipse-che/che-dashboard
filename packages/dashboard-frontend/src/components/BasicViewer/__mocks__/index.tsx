/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/components/BasicViewer';

export class BasicViewer extends React.PureComponent<Props> {
  public render(): React.ReactNode {
    const { id, value } = this.props;

    return (
      <div>
        Mock Basic Viewer
        <span data-testid="basic-viewer-id">{id}</span>
        <span data-testid="basic-viewer-value">{value}</span>
      </div>
    );
  }
}
