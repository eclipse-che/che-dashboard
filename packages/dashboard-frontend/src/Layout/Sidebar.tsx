/*
 * Copyright (c) 2018-2023 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { PageSidebar } from '@patternfly/react-core';
import { History } from 'history';

import Navigation from './Navigation';

type Props = {
  isManaged: boolean;
  isNavOpen: boolean;
  history: History;
};

export default class Sidebar extends React.PureComponent<Props> {
  public render(): React.ReactElement {
    const { isNavOpen, history } = this.props;

    return <PageSidebar isNavOpen={isNavOpen} nav={<Navigation history={history} />} />;
  }
}
