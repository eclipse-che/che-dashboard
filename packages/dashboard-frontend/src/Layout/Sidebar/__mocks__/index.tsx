/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/Layout/Sidebar';

export class Sidebar extends React.PureComponent<Props> {
  public render(): React.ReactElement {
    return (
      <div>
        Mock Sidebar component
        <span data-testid="isManaged">{this.props.isManaged ? 'true' : 'false'}</span>
        <span data-testid="isNavOpen">{this.props.isNavOpen ? 'true' : 'false'}</span>
      </div>
    );
  }
}
