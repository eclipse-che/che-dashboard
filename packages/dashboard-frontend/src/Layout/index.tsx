/*
 * Copyright (c) 2018-2023 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { connect, ConnectedProps } from 'react-redux';
import { Brand, Page } from '@patternfly/react-core';
import { History } from 'history';
import { matchPath } from 'react-router';
import Header from './Header';
import Sidebar from './Sidebar';
import StoreErrorsAlert from './StoreErrorsAlert';
import { AppState } from '../store';
import { lazyInject } from '../inversify.config';
import { IssuesReporterService } from '../services/bootstrap/issuesReporter';
import { ErrorReporter } from './ErrorReporter';
import { IssueComponent } from './ErrorReporter/Issue';
import { BannerAlert } from '../components/BannerAlert';
import { ErrorBoundary } from './ErrorBoundary';
import { ROUTE } from '../Routes/routes';
import { selectBranding } from '../store/Branding/selectors';
import { ToggleBarsContext } from '../contexts/ToggleBars';
import { signOut } from '../services/helpers/login';
import { selectDashboardLogo } from '../store/ServerConfig/selectors';

const IS_MANAGED_SIDEBAR = false;

type Props = MappedProps & {
  children: React.ReactNode;
  history: History;
};
type State = {
  isSidebarVisible: boolean;
  isHeaderVisible: boolean;
};

export class Layout extends React.PureComponent<Props, State> {
  @lazyInject(IssuesReporterService)
  private readonly issuesReporterService: IssuesReporterService;

  constructor(props: Props) {
    super(props);

    this.state = {
      isHeaderVisible: true,
      isSidebarVisible: true,
    };
  }

  private toggleNav(): void {
    this.setState({
      isSidebarVisible: !this.state.isSidebarVisible,
    });
  }

  private hideAllBars(): void {
    this.setState({
      isHeaderVisible: false,
      isSidebarVisible: false,
    });
  }

  private showAllBars(): void {
    this.setState({
      isHeaderVisible: true,
      isSidebarVisible: true,
    });
  }

  public componentDidMount(): void {
    const matchFactoryLoaderPath = matchPath(this.props.history.location.pathname, {
      path: ROUTE.FACTORY_LOADER,
    });
    const matchIdeLoaderPath = matchPath(this.props.history.location.pathname, {
      path: ROUTE.IDE_LOADER,
    });
    if (matchFactoryLoaderPath !== null || matchIdeLoaderPath !== null) {
      this.hideAllBars();
    }
  }

  public render(): React.ReactElement {
    /* check for startup issues */
    if (this.issuesReporterService.hasIssue) {
      const issue = this.issuesReporterService.reportIssue();
      const brandingData = this.props.branding;
      if (issue) {
        return (
          <ErrorReporter>
            <IssueComponent branding={brandingData} issue={issue} />
          </ErrorReporter>
        );
      }
    }

    const { isHeaderVisible, isSidebarVisible } = this.state;
    const { history, branding, dashboardLogo } = this.props;

    const logoSrc =
      dashboardLogo !== undefined
        ? `data:${dashboardLogo.mediatype};base64,${dashboardLogo.base64data}`
        : branding.logoFile;

    return (
      <ToggleBarsContext.Provider
        value={{
          hideAll: () => this.hideAllBars(),
          showAll: () => this.showAllBars(),
        }}
      >
        <Page
          header={
            <Header
              history={history}
              isVisible={isHeaderVisible}
              logo={<Brand src={logoSrc} alt="Logo" />}
              logout={() => signOut()}
              toggleNav={() => this.toggleNav()}
            />
          }
          sidebar={
            <Sidebar
              isManaged={IS_MANAGED_SIDEBAR}
              isNavOpen={isSidebarVisible}
              history={history}
            />
          }
          isManagedSidebar={IS_MANAGED_SIDEBAR}
        >
          <ErrorBoundary>
            <StoreErrorsAlert />
            <BannerAlert />
            {this.props.children}
          </ErrorBoundary>
        </Page>
      </ToggleBarsContext.Provider>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  branding: selectBranding(state),
  dashboardLogo: selectDashboardLogo(state),
});

const connector = connect(mapStateToProps);

type MappedProps = ConnectedProps<typeof connector>;
export default connector(Layout);
