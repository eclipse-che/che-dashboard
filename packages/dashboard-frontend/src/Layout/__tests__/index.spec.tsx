/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { Store } from 'redux';

import { container } from '@/inversify.config';
import Layout from '@/Layout';
import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';
import { BrandingData } from '@/services/bootstrap/branding.constant';
import { IssuesReporterService } from '@/services/bootstrap/issuesReporter';
import { FakeStoreBuilder } from '@/store/__mocks__/storeBuilder';
import * as SanityCheckStore from '@/store/SanityCheck';

const issuesReporterService = container.get(IssuesReporterService);

jest.mock('@/Layout/ErrorBoundary');
jest.mock('@/Layout/Header');
jest.mock('@/Layout/Sidebar');
jest.mock('@/Layout/StoreErrorsAlert');
jest.mock('@/components/BannerAlert');
jest.mock('@/contexts/ToggleBars');

const mockLogout = jest.fn().mockResolvedValue(undefined);
jest.mock('@/services/helpers/login.ts', () => ({
  signOut: () => mockLogout(),
}));

const mockTestBackends = jest.fn();
jest.mock('@/store/SanityCheck/index', () => ({
  ...jest.requireActual('@/store/SanityCheck/index'),
  actionCreators: {
    testBackends: () => async () => mockTestBackends(),
  } as SanityCheckStore.ActionCreators,
}));

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

describe('Layout component', () => {
  let store: Store;

  beforeEach(() => {
    store = new FakeStoreBuilder()
      .withBranding({
        logoFile: 'logo-File',
      } as BrandingData)
      .build();
  });

  afterEach(() => {
    issuesReporterService.clearIssues();
    jest.clearAllMocks();
  });

  test('snapshot of the namespace provisioning error page', () => {
    issuesReporterService.registerIssue(
      'namespaceProvisioningError',
      new Error('Namespace provisioning error'),
    );
    const snapshot = createSnapshot(store);
    expect(snapshot).toMatchSnapshot();
  });

  test('snapshot of the default page', () => {
    const snapshot = createSnapshot(store);
    expect(snapshot).toMatchSnapshot();
  });

  test('logout', async () => {
    renderComponent(store);
    const logoutButton = screen.getByRole('button', { name: 'logout' });
    await userEvent.click(logoutButton);

    expect(mockLogout).toHaveBeenCalledTimes(1);
  });

  test('toggleNav', async () => {
    renderComponent(store);

    expect(screen.getByTestId('isNavOpen')).toHaveTextContent('true');

    const toggleNavButton = screen.getByRole('button', { name: 'toggleNav' });
    await userEvent.click(toggleNavButton);

    expect(screen.getByTestId('isNavOpen')).toHaveTextContent('false');
  });

  test('onError', async () => {
    mockTestBackends.mockResolvedValue(undefined);

    renderComponent(store);

    const btn = screen.getByRole('button', { name: 'onError' });
    await userEvent.click(btn);

    expect(mockTestBackends).toHaveBeenCalled();
  });

  test('onNamespaceProvisioningError', async () => {
    const errorMessage = 'Namespace provisioning error';
    issuesReporterService.registerIssue('namespaceProvisioningError', new Error(errorMessage));

    renderComponent(store);

    const errorMessageContainer = await waitFor(() => screen.queryByText(errorMessage));

    expect(errorMessageContainer).not.toBeNull();
  });
});

function getComponent(store: Store) {
  const history = createMemoryHistory();
  return (
    <Provider store={store}>
      <Layout history={history}>
        <div>Test</div>
      </Layout>
    </Provider>
  );
}
