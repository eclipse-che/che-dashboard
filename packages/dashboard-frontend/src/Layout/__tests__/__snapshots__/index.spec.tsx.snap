// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout component snapshot 1`] = `
<div
  className="pf-c-page"
>
  <div>
    Mock Header component
    <button
      onClick={[Function]}
    >
      toggleNav
    </button>
    <button
      onClick={[Function]}
    >
      logout
    </button>
  </div>
  <div>
    Mock Sidebar component
    <span
      data-testid="isManaged"
    >
      false
    </span>
    <span
      data-testid="isNavOpen"
    >
      true
    </span>
  </div>
  <main
    className="pf-c-page__main"
    tabIndex={-1}
  >
    <div>
      Mock ErrorBoundary component
      <button
        onClick={[Function]}
      >
        onError
      </button>
      <div>
        Mock StoreErrorsAlert component
      </div>
      <div>
        Mock BannerAlert component
      </div>
      <div>
        Test
      </div>
    </div>
  </main>
</div>
`;
